loader/lightning.go

```diff
@@ -0,0 +1,273 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"context"
+	"path/filepath"
+	"sync"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/dm/unit"
+	"github.com/pingcap/dm/pkg/binlog"
+	"github.com/pingcap/dm/pkg/conn"
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/utils"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/br/pkg/lightning"
+	lcfg "github.com/pingcap/tidb/br/pkg/lightning/config"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+const (
+	lightningCheckpointFile  = "lightning.checkpoint.0.sql"
+	lightningCheckpointDB    = "lightning"
+	lightningCheckpointTable = "checkpoint"
+)
+
+// LightningLoader can load your mydumper data into TiDB database.
+type LightningLoader struct {
+	sync.RWMutex
+
+	cfg             *config.SubTaskConfig
+	cli             *clientv3.Client
+	checkPoint      CheckPoint
+	workerName      string
+	logger          log.Logger
+	core            *lightning.Lightning
+	toDB            *conn.BaseDB
+	toDBConns       []*DBConn
+	lightningConfig *lcfg.GlobalConfig
+
+	finish atomic.Bool
+	closed atomic.Bool
+	cancel context.CancelFunc // for per task context, which maybe different from lightning context
+}
+
+// NewLightning creates a new Loader importing data with lightning.
+func NewLightning(cfg *config.SubTaskConfig, cli *clientv3.Client, workerName string) *LightningLoader {
+	lightningCfg := makeGlobalConfig(cfg)
+	core := lightning.New(lightningCfg)
+	loader := &LightningLoader{
+		cfg:             cfg,
+		cli:             cli,
+		core:            core,
+		lightningConfig: lightningCfg,
+		logger:          log.With(zap.String("task", cfg.Name), zap.String("unit", "lightning-load")),
+		workerName:      workerName,
+	}
+	return loader
+}
+
+func makeGlobalConfig(cfg *config.SubTaskConfig) *lcfg.GlobalConfig {
+	lightningCfg := lcfg.NewGlobalConfig()
+	if cfg.To.Security != nil {
+		lightningCfg.Security.CAPath = cfg.To.Security.SSLCA
+		lightningCfg.Security.CertPath = cfg.To.Security.SSLCert
+		lightningCfg.Security.KeyPath = cfg.To.Security.SSLKey
+	}
+	lightningCfg.TiDB.Host = cfg.To.Host
+	lightningCfg.TiDB.Psw = cfg.To.Password
+	lightningCfg.TiDB.User = cfg.To.User
+	lightningCfg.TiDB.Port = cfg.To.Port
+	lightningCfg.TiDB.StatusPort = cfg.TiDB.StatusPort
+	lightningCfg.TiDB.PdAddr = cfg.TiDB.PdAddr
+	lightningCfg.TikvImporter.Backend = cfg.TiDB.Backend
+	lightningCfg.PostRestore.Checksum = lcfg.OpLevelOff
```

Comment:
will we need it for local backend?

Reply:
But dm may load multiple sharding to tidb, and we can not guarentee that no other client try to write data to tidb during checksum

Reply:
Actually tidb-backend doesn't support any kind of `checksum`😅